### PR TITLE
[Feat]: Migrate NFT Actions, Mutations and State

### PR DESCRIPTION
**Pull request by Jennievon**

### CODE REVIEW

File: src/store/actions/getNFTAssets.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        ? true
+        : false
+    } else {
+      asset['starred'] = false
+    }
+  })
+
+  commit('SET_NFT_ASSETS_NUMBER', nft.assets.length)
+
+  let result = nft.assets.reduce(function (r, a) {
```
Comment (monokh) [inline src/store/actions/getNFTAssets.js]: Name `r` and `a` so it's clearer what's happening

File: src/store/types.ts [hunk 2]
```diff
@@ -208,4 +208,126 @@ export interface RootState {
   experiments: Partial<Record<ExperimentType, boolean>>;
   watsNewModalVersion: string;
   enabledChains: WalletIdNetworkMap<ChainId[]>;
+
+  nftAssets: NFTAssets,
+  nftAssetsNumber: number,
+  starredNFTs: NFTAsset[]
+}
+
+export type NFTAsset = {
```
Comment (monokh) [inline src/store/types.ts]: Why is this type defined in wallet-core? Doesn't it come from CAL/chainify? It should be imported from there

File: src/store/actions/getNFTAssets.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+          item.asset_contract.address === asset.asset_contract.address && item.id === asset.id
+      ).length
+        ? true
+        : false
+    } else {
+      asset['starred'] = false
+    }
+  })
+
+  commit('SET_NFT_ASSETS_NUMBER', nft.assets.length)
```
Comment (monokh) [inline src/store/actions/getNFTAssets.js]: There's no need to commit something to the state that can be derived from the state directly.
When you need the length you can do state.nfts.length

File: src/store/actions/getNFTAssets.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+  let result = nft.assets.reduce(function (r, a) {
+    r[a.collection.name] = r[a.collection.name] || []
```
Comment (monokh) [inline src/store/actions/getNFTAssets.js]: Sorting and grouping should not be in the action and is more of a UI concern.
Rather have all NFTs flat in the structure and sorting/ grouping happening elsewhere. You can add it as a getter if you want - there are similar ones in the getters you can look at

File: src/store/state.ts [hunk 5]
```diff
@@ -49,6 +49,9 @@ const initialState: RootState = {
   experiments: {},
   watsNewModalVersion: '',
   enabledChains: {},
+  nftAssets: {},
+  nftAssetsNumber: 0,
+  starredNFTs: []
```
Comment (monokh) [inline src/store/state.ts]: There seems to be a lot of complexity introduced because of this array which is separately storing the starred NFTs. It would be better if `starred` was just a property of the NFT, to get a list of starred NFTs, you have to just filter that and that can be done on the UI

File: src/store/mutations.ts [hunk 6]
```diff
@@ -423,6 +425,34 @@ export default {
       ...payload,
     };
   },
+  SET_NFT_ASSETS(state: RootState, payload: NFTAssets) {
+    state.nftAssets = payload
```
Comment (monokh) [inline src/store/mutations.ts]: Aren't the nfts different depending on the network and walletid at the very least? what about depending on account? If so, then they need to be split in the state under those categories. See for example the `history` state object

**Review by monokh — CHANGES_REQUESTED** [review]
Please also add a test for the actions that does some basic checks that nfts are retrieved

File: src/factory/client.ts [hunk 7]
```diff
@@ -123,7 +124,7 @@ function createEthereumClient(
   } else {
     ethClient.addProvider(
       new EthereumJsWalletProvider({
-        network: ethereumNetwork,
+        network: { ...ethereumNetwork, chainId: 4 },
```
Comment (monokh) [inline src/factory/client.ts]: Are we really switching to rinkeby? As mentioned this will break thorchain

File: src/factory/client.ts [hunk 8]
```diff
@@ -147,6 +148,9 @@ function createEthereumClient(
   // TODO: remove when Chainify is added
   (ethClient.wallet as any).signTypedMessage = signTypedMessage.bind(ethClient.wallet);
 
+  const openSeaAPI = 'https://rinkeby-api.opensea.io/api/v1/';
```
Comment (monokh) [inline src/factory/client.ts]: Only rinkeby? What about mainnet?

File: src/factory/client.ts (hunk 7 above)
Comment (monokh) [reply src/factory/client.ts]: maybe you can share the rinkeby version with QA just for testing, but keep ropsten in the official build

File: src/store/actions/performNextAction/sendNFT.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    network,
+    walletId,
+    asset: transaction.from,
+    accountId: transaction.accountId,
+  });
+  try {
+    const tx = await client.chain.getTransactionByHash(transaction.txHash);
+    console.log('🚀 ~ file: sendNFT.ts ~ line 30 ~ tx', tx);
+    if (tx && tx.confirmations && tx.confirmations > 0) {
+      dispatch.updateBalances({
```
Comment (monokh) [inline src/store/actions/performNextAction/sendNFT.ts]: Why do you need to update balances? Isn't that just for assets not NFTs?

Comment (monokh) [reply src/store/actions/performNextAction/sendNFT.ts]: Should you update the NFTs of the account?

File: src/store/actions/sendNFTTransaction.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    fee,
+    feeLabel,
+    fiatRate,
+    nft,
+  }: // data,
+  {
+    network: Network;
+    accountId: AccountId;
+    walletId: WalletId;
+    contract: Address | string;
```
Comment (monokh) [inline src/store/actions/sendNFTTransaction.ts]: Shouldn't these values "contract", "Receiver" "tokenIDs" all come from the `nft` ? Having duplicates of the data can cause problems

File: src/store/getters.ts [hunk 11]
```diff
@@ -282,4 +282,17 @@ export default {
     }
     return false;
   },
+  nftAssetsByCollection(...context: GetterContext): NFTAsset[] {
+    const { state } = rootGetterContext(context);
+    const { nftAssets } = state;
+    const result = nftAssets.reduce(function (assets: any, asset: any) {
```
Comment (monokh) [inline src/store/getters.ts]: No `any` please, the type of these objects should be known

File: src/store/getters.ts [hunk 12]
```diff
@@ -282,4 +282,17 @@ export default {
...
+      assets[asset.collection.name] = assets[asset.collection.name] || [];
+      assets[asset.collection.name].push(asset);
+      assets[asset.collection.name].sort(function (assetA: any, assetB: any) {
+        return assetA.starred === assetB.starred ? 0 : assetA.starred ? -1 : 1;
+      });
+      return assets;
+    }, Object.create(null));
```
Comment (monokh) [inline src/store/getters.ts]: `Object.create(null)` vs `{}`?

File: src/store/types.ts [hunk 13]
```diff
@@ -135,6 +136,18 @@ export interface SendHistoryItem extends BaseHistoryItem {
...
 
+export interface NFTSendHistoryItem extends BaseHistoryItem {
+  type: TransactionType.NFT;
+  toAddress: string;
+  amount: string;
+  tx: Transaction;
+  nft: NFTAsset;
+  txHash: string;
+  accountId: AccountId;
+  fiatRate: number;
```
Comment (monokh) [inline src/store/types.ts]: is this needed?

File: src/store/types.ts [hunk 14]
```diff
@@ -135,6 +136,18 @@ export interface SendHistoryItem extends BaseHistoryItem {
   status: SendStatus;
 }
 
+export interface NFTSendHistoryItem extends BaseHistoryItem {
+  type: TransactionType.NFT;
+  toAddress: string;
+  amount: string;
```
Comment (monokh) [inline src/store/types.ts]: does NFT have an amount?

File: src/store/state.ts [hunk 15]
```diff
@@ -49,6 +49,8 @@ const initialState: RootState = {
   experiments: {},
   watsNewModalVersion: '',
   enabledChains: {},
+  nftAssets: [],
+  starredNFTs: [],
```
Comment (monokh) [inline src/store/state.ts]: I left a comment about not requiring a separate state for starred NFTs and just using the property in `nftAssets`. You can again use a getter if you really need it in a list like this

File: src/store/mutations.ts [hunk 16]
```diff
@@ -427,6 +437,23 @@ export default {
       ...payload,
     };
   },
+  SET_NFT_ASSETS(state: RootState, payload: { nftAssets: NFTAsset[]; network: Network; walletId: WalletId }) {
+    // const walletHistory = state.history[payload.network]?.[payload.walletId];
+    state.nftAssets = payload.nftAssets;
```
Comment (monokh) [inline src/store/mutations.ts]: Don't think this comment has been addressed yet: 
https://github.com/liquality/wallet-core/pull/6#discussion_r867830110

We have to figure out handling different chain/network/account now. Once the state is created, it will be difficult to migrate users to different state when we need support for it (which would be very soon)

**Review by monokh — CHANGES_REQUESTED** [review]

File: src/factory/client.ts (hunk 7 above)
Comment (Jennievon) [reply src/factory/client.ts]: How do we handle it when we need to release though?

File: src/store/types.ts (hunk 13 above)
Comment (Jennievon) [reply src/store/types.ts]: By 'this', do you mean `fiatRate`?

File: src/store/actions/sendNFTTransaction.ts (hunk 10 above)
Comment (Jennievon) [reply src/store/actions/sendNFTTransaction.ts]: Yeah, just the `contract` and `tokenIDs` though but noted. 👌🏽

File: src/store/types.ts (hunk 13 above)
Comment (monokh) [reply src/store/types.ts]: yes

File: src/factory/client.ts (hunk 7 above)
Comment (monokh) [reply src/factory/client.ts]: I'm not sure we can support testnet NFTs with the current state of the wallet. The liquidity providers, liquality and thorchain are running on ropsten and we can't sacrifice that for NFTs

We have to have a separate conversation for how to enable multiple testnet networks for the wallet

Comment (AndonMitev) [reply src/factory/client.ts]: I heard somewhere rinkeby will be deprecated

File: src/store/actions/getNFTAssets.ts [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+  }: {
+    network: Network;
+    walletId: WalletId;
+  }
+): Promise<any> => {
+  const { state, commit, getters } = rootActionContext(context);
+  const client = getters.client({
+    network,
+    walletId,
+    asset: 'ETH',
```
Comment (AndonMitev) [inline src/store/actions/getNFTAssets.ts]: What if we want to get for polygon for example

File: src/store/actions/getNFTAssets.ts [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+    if (state.starredNFTs) {
+      asset['starred'] = state.starredNFTs.filter(
+        (item) => item.asset_contract.address === asset.asset_contract.address && item.id === asset.id
+      ).length
+        ? true
+        : false;
+      nftAssets.push(asset);
+    } else {
+      asset['starred'] = false;
+      nftAssets.push(asset);
```
Comment (AndonMitev) [inline src/store/actions/getNFTAssets.ts]: you can move `nftAssets.push` outside the if else because it's executed anytime anyway

File: src/store/actions/getNFTAssets.ts [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    }
+  });
+  console.log('🚀 ~ file: getNFTAssets.ts ~ line 34 ~ nftAssets ~ nftAssets', nftAssets);
```
Comment (AndonMitev) [inline src/store/actions/getNFTAssets.ts]: `console.log`

File: src/store/actions/performNextAction/sendNFT.ts [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+  const { getters } = rootActionContext(context);
+  const client = getters.client({
+    network,
+    walletId,
+    asset: transaction.from,
+    accountId: transaction.accountId,
+  });
+  try {
+    const tx = await client.chain.getTransactionByHash(transaction.txHash);
+    console.log('🚀 ~ file: sendNFT.ts ~ line 30 ~ tx', tx);
```
Comment (AndonMitev) [inline src/store/actions/performNextAction/sendNFT.ts]: log

**Review by bradleySuira — APPROVED** [review]
LGTM

File: src/store/actions/getNFTAssets.ts [hunk 21]
```diff
@@ -22,20 +22,50 @@ export const getNFTAssets = async (
...
-  const nftAssets: NFTAsset[] = [];
-  nft.assets.forEach((asset: any) => {
-    if (state.starredNFTs) {
-      asset['starred'] = state.starredNFTs.filter(
-        (item) => item.asset_contract.address === asset.asset_contract.address && item.id === asset.id
-      ).length
-        ? true
-        : false;
-      nftAssets.push(asset);
+  const oldAssets: NFTAsset[] = state.nftAssets;
```
Comment (monokh) [inline src/store/actions/getNFTAssets.ts]: What is the purpose of this function?  Can you describe input and outputs?
It's really hard to understand and I feel there is a much clearer way to write what this is doing

**Review by monokh — CHANGES_REQUESTED** [review]

File: src/store/actions/getNFTAssets.ts [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+  });
+
+  const nft = await client.nft.fetch();
+
+  const account =  state.accounts[walletId]![network].find((a) => a.id === accountId);
+
+  const nftAssetsStoredInState: NFTAsset[] = account?.nftAssets || [];
+  const nftAssetsFetched: NFTAsset[] = nft.assets;
+
+  // set new nft assets to a new map
```
Comment (monokh) [inline src/store/actions/getNFTAssets.ts]: I see that you added comments but we really need to simplify this function overall
It looks like it's jsut merging 2 arrays and default starred = false. That shouldn't require this much code and should be doable with some map+reduce. Can you help @kraikov ?

Comment (kraikov) [reply src/store/actions/getNFTAssets.ts]: I think we can simply use the fetched array and set the starred ones from the state.

Comment (Jennievon) [reply src/store/actions/getNFTAssets.ts]: Thanks, @monokh @kraikov. 👍 
This has been addressed.

File: src/factory/client.ts (hunk 7 above)
Comment (kraikov) [reply src/factory/client.ts]: We should not switch to Rinkeby. Can you please revert this back.

File: src/store/actions/getNFTAssets.ts [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+  const nft = await client.nft.fetch();
+
+  const account =  state.accounts[walletId]![network].find((a) => a.id === accountId);
+
+  const nftAssetsStoredInState: NFTAsset[] = account?.nftAssets || [];
+  const nftAssetsFetched: NFTAsset[] = nft.assets;
+
+  const nftAssets: NFTAsset[] = [];
+
+  nftAssetsFetched.forEach((nftAsset) => {
```
Comment (kraikov) [inline src/store/actions/getNFTAssets.ts]: can you use `map` instead of `forEach`?

File: src/store/actions/sendNFTTransaction.ts [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+    accountId: AccountId;
+    walletId: WalletId;
+    receiver: string;
+    values: number[];
+    fee: number;
+    feeLabel: FeeLabel;
+    nft: NFTAsset;
+  }
+): Promise<any> => {
+  const asset = 'ETH';
```
Comment (kraikov) [inline src/store/actions/sendNFTTransaction.ts]: this one is confusing.... sending an NFT should not contain any `from` and `to` assets. Can we set them to `null` ?

File: src/store/actions/sendNFTTransaction.ts [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+    network,
+    accountId,
+    walletId,
+    receiver,
+    values,
+    fee,
+    feeLabel,
+    nft,
+  }: {
+    network: Network;
```
Comment (kraikov) [inline src/store/actions/sendNFTTransaction.ts]: Can we please have a dedicated type for the input param?

File: src/store/actions/sendNFTTransaction.ts [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+    accountId: AccountId;
+    walletId: WalletId;
+    receiver: string;
+    values: number[];
+    fee: number;
+    feeLabel: FeeLabel;
+    nft: NFTAsset;
+  }
+): Promise<any> => {
```
Comment (kraikov) [inline src/store/actions/sendNFTTransaction.ts]: Avoid using `any`

File: src/store/getters.ts [hunk 27]
```diff
@@ -282,4 +292,36 @@ export default {
...
     return false;
   },
+  nftAssetsByCollection(...context: GetterContext): NFTAsset[] {
+    const { getters } = rootGetterContext(context);
+    const { nftAssetsByAccount } = getters;
+    let nftAssetsByCollection: NFTAsset[] = [];
+    // spread all nfts into one array
+    Object.values(nftAssetsByAccount).forEach((nftAssets) => {
+      nftAssetsByCollection = {
+        ...nftAssetsByCollection,
```
Comment (kraikov) [inline src/store/getters.ts]: `nftAssetsByCollection` is always an empty array, right?

File: src/store/getters.ts [hunk 28]
```diff
@@ -282,4 +292,36 @@ export default {
...
+    accountsData.forEach((account) => {
+      if(account.nftAssets) {
+       const result = account.nftAssets.reduce(function (assets: NFTAssets, asset: NFTAsset) {
+            assets[asset.collection.name] = assets[asset.collection.name] || [];
+            assets[asset.collection.name].push(asset);
+            assets[asset.collection.name].sort(function (assetA: NFTAsset, assetB: NFTAsset) {
+              return assetA.starred === assetB.starred ? 0 : assetA.starred ? -1 : 1;
+            });
+            return assets;
+          }, Object.create({}));
```
Comment (kraikov) [inline src/store/getters.ts]: can't you just use `{}`

File: src/store/getters.ts [hunk 29]
```diff
@@ -282,4 +292,36 @@ export default {
...
+  },
+  nftAssetsByAccount(...context: GetterContext): NFTAssets {
+    const { getters } = rootGetterContext(context);
+    const { accountsData } = getters;
+    const nftAssetsByAccount: NFTAssets = {};
+    accountsData.forEach((account) => {
+      if(account.nftAssets) {
+       const result = account.nftAssets.reduce(function (assets: NFTAssets, asset: NFTAsset) {
+            assets[asset.collection.name] = assets[asset.collection.name] || [];
+            assets[asset.collection.name].push(asset);
```
Comment (kraikov) [inline src/store/getters.ts]: the above 2 lines can be combiled with:

`(assets[asset.collection.name] ||= []).push(asset)`

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR_assignment

**Review by kraikov — CHANGES_REQUESTED** [review]
`package-lock.json` must not be included, because wallet-core uses `yarn.lock`

File: src/store/actions/sendNFTTransaction.ts [hunk 30]
```diff
@@ -0,0 +1,80 @@
...
+    fee,
+    tx,
+    nft,
+    txHash: tx.hash,
+    startTime: Date.now(),
+    status: SendStatus.WAITING_FOR_CONFIRMATIONS,
+    accountId,
+    feeLabel,
+  };
+  console.log('🚀 ~ file: sendNFTTransaction.ts ~ line 64 ~ transaction', transaction);
```
Comment (kraikov) [inline src/store/actions/sendNFTTransaction.ts]: Remove.

**Review by kraikov — CHANGES_REQUESTED** [review]

File: src/store/getters.ts (hunk 27 above)
Comment (Jennievon) [reply src/store/getters.ts]: For now but when we have NFTs in different accounts, we would still need to spread them into the array.

Comment (kraikov) [reply src/store/getters.ts]: oh.. got it... then it will be better to use `reduce` and use different name for the function and parameter because it's a little bit confusing

File: src/store/actions/sendNFTTransaction.ts (hunk 24 above)
Comment (kraikov) [reply src/store/actions/sendNFTTransaction.ts]: can we pass the asset as function param, because in the future we would have different NFTs (e.g. Solana)

Comment (kraikov) [reply src/store/actions/sendNFTTransaction.ts:25]: can we pass the asset as function param, because in the future we would have different NFTs (e.g. Solana)

File: src/store/actions/getNFTAssets.ts [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+  const nftAssetsFetched: NFTAsset[] = nft.assets;
+
+  const nftAssets: NFTAsset[] = [];
+
+  nftAssetsFetched.map((nftAsset) => {
+    const nftAssetStoredInState = nftAssetsStoredInState.find((a) => a.id === nftAsset.id);
+    if (nftAssetStoredInState) {
+      nftAsset.starred = nftAssetStoredInState.starred;
+    }
+    nftAssets.push(nftAsset);
```
Comment (kraikov) [inline src/store/actions/getNFTAssets.ts]: the map function populates a new array with the returned value, so there is no need to do 
`nftAssets.push(nftAsset);`

You should do:

`const nftAssets =  nftAssetsFetched.map((nftAsset) => .... return nftAsset)`

Comment (kraikov) [inline src/store/actions/getNFTAssets.ts:39]: the map function populates a new array with the returned value, so there is no need to do 
`nftAssets.push(nftAsset);`

You should do:

`const nftAssets =  nftAssetsFetched.map((nftAsset) => .... return nftAsset)`

File: src/store/getters.ts [hunk 32]
```diff
@@ -298,4 +308,34 @@ export default {
     }
     return false;
   },
+  nftAssetsByCollection(...context: GetterContext): NFTAsset[] {
+    const { getters } = rootGetterContext(context);
+    const { nftAssetsByAccount } = getters;
+    let nftAssetsByCollection: NFTAsset[] = [];
+    nftAssetsByCollection = Object.values(nftAssetsByAccount).reduce((accum, nftAssets) => {
```
Comment (kraikov) [inline src/store/getters.ts]: The `reduce` function is not used correctly here. Please check https://codeburst.io/how-to-reduce-array-to-a-single-type-11cfe2bcae66

Comment (kraikov) [inline src/store/getters.ts]: The `reduce` function is not used correctly here. Please check https://codeburst.io/how-to-reduce-array-to-a-single-type-11cfe2bcae66

File: src/store/getters.ts (hunk 28 above)
Comment (kraikov) [reply src/store/getters.ts:335]: This is not yet resolved

Comment (kraikov) [reply src/store/getters.ts]: This is not yet resolved

File: src/factory/client/clients.ts [hunk 33]
```diff
@@ -87,11 +87,14 @@ export function createEVMClient(
...
   mnemonic: string,
   accountType: AccountType,
-  derivationPath: string
+  derivationPath: string,
+  nftProvider?: EvmNftProvider
 ) {
   // disable multicall for Testnets
   const chainProvider = new EvmChainProvider(ethereumNetwork, undefined, feeProvider, !ethereumNetwork.isTestnet);
   const swapProvider = new EvmSwapProvider({ contractAddress: HTLC_CONTRACT_ADDRESS });
+  const walletOptions = { derivationPath, mnemonic };
```
Comment (monokh) [inline src/factory/client/clients.ts]: Unrelated change?

File: src/factory/client/evm.ts [hunk 34]
```diff
@@ -1,25 +1,33 @@
...
+    responseType: 'text',
+    transformResponse: undefined,
+    headers: {
+      'X-Api-Key': '',
+    },
+  };
   const feeProvider = new EIP1559FeeProvider(ethNetwork.rpcUrl as string);
-  return createEVMClient(ethNetwork, feeProvider, mnemonic, accountType, derivationPath);
+  const chainProvider = new EvmChainProvider(
+    { ...ethNetwork, chainId: 4 },
```
Comment (monokh) [inline src/factory/client/evm.ts]: As mentioned before, please let the wallet stay on ropsten testnet

File: src/store/actions/getNFTAssets.ts [hunk 35]
```diff
@@ -0,0 +1,47 @@
...
+  {
+    walletId,
+  }: {
+    walletId: WalletId;
+  }
+): Promise<void> => {
+  const { state, commit, getters } = rootActionContext(context);
+  const { networks, nftAssets } = buildConfig;
+
+  networks.forEach((network) => {
```
Comment (monokh) [inline src/store/actions/getNFTAssets.ts]: We shouldn't update for all networks when the user is not on it. It will slow down the wallet for no reason, 99% only use mainnet. the active network should be passed into this function

File: src/store/actions/getNFTAssets.ts [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+    const assetKeys = nftAssets[network];
```
Comment (monokh) [inline src/store/actions/getNFTAssets.ts]: assetKeys? Isn't it chains?

File: src/store/actions/getNFTAssets.ts [hunk 37]
```diff
@@ -0,0 +1,47 @@
...
+    assetKeys.forEach(async (chain) => {
+      const asset = chains[chain].nativeAsset;
+      const account = state.accounts[walletId]![network].find((a) => a.assets.includes(asset));
```
Comment (monokh) [inline src/store/actions/getNFTAssets.ts]: Why do you need to find the account by the asset? You are looking for the account for that chain and accounts have a `chain` property

Comment (monokh) [reply src/store/actions/getNFTAssets.ts]: There can be multiple accounts for a chain, each account (therefore address) has it's own list of nfts. You need to make sure every account is processed

File: src/store/actions/getNFTAssets.ts [hunk 38]
```diff
@@ -0,0 +1,47 @@
...
+      const _client = getters.client({
+        network,
+        walletId,
+        asset,
+        accountId: account?.id,
+      });
+
+      const nft: any = await _client.nft?.fetch();
```
Comment (monokh) [inline src/store/actions/getNFTAssets.ts]: why do we type this as any? I believe it does return a type

File: src/store/actions/getNFTAssets.ts [hunk 39]
```diff
@@ -0,0 +1,47 @@
...
+
+      const nftAssetsStoredInState: NFTAsset[] = account?.nftAssets || [];
+      const nftAssetsFetched: NFTAsset[] = nft.assets;
+
+      const nftAssets: NFTAsset[] = nftAssetsFetched.map((nftAsset) => {
+        const nftAssetStoredInState = nftAssetsStoredInState.find((asset) => asset.id === nftAsset.id);
+        if (nftAssetStoredInState) {
+          nftAsset.starred = nftAssetStoredInState.starred || false;
+          nftAsset.asset = nftAssetStoredInState.asset || asset;
```
Comment (monokh) [inline src/store/actions/getNFTAssets.ts]: These 2 statements seem weird, asset and chain are static aren't they? they should always be the same so why the  || ?
Also, why store asset? an nft is for a chain, doesn't have anything to do with an asset i believe

File: src/store/migrations/before_after_migration_script.test.ts [hunk 40]
```diff
@@ -0,0 +1,116 @@
+import * as Process from 'process';
```
Comment (monokh) [inline src/store/migrations/before_after_migration_script.test.ts]: unrelated?

File: src/store/types.ts (hunk 2 above)
Comment (monokh) [reply src/store/types.ts]: ?

**Review by monokh — CHANGES_REQUESTED** [review]

File: src/store/actions/getNFTAssets.ts (hunk 38 above)
Comment (Jennievon) [reply src/store/actions/getNFTAssets.ts]: `void` is being set as the type in `chainify` for the `fetch` response, hence, we cannot set a specific type as it isn't assignable to `void`. That's the reason I used `any`

File: src/factory/client/evm.ts (hunk 34 above)
Comment (Jennievon) [reply src/factory/client/evm.ts]: This wouldn't be merged. It's just to test the NFTs on testnet.

File: src/store/migrations/before_after_migration_script.test.ts (hunk 40 above)
Comment (Jennievon) [reply src/store/migrations/before_after_migration_script.test.ts]: some file changes are from the `master` branch

Comment (monokh) [reply src/store/migrations/before_after_migration_script.test.ts]: does that mean chainify branch needs to be updated to include something? This PR should not be adding anything unrelated

File: src/store/actions/getNFTAssets.ts (hunk 38 above)
Comment (monokh) [reply src/store/actions/getNFTAssets.ts]: Yea it's not good that chainify is returning any 
Is the actual type `NftAsset` that you have in this project? In that case we can move it to chanify so you get the type here

**Comment by monokh** [comment]
Closing to reopen because github actions behaving funny
